### DISCUSSION TITLE
RFC: Property per cmd line flag vs single ExtraArgs property in Custom Resources

### DISCUSSION THREAD

**Discussion by arajkumar** [prometheus-operator/prometheus-operator#4109] [Ideas]
Currently almost all custom resources from Prometheus operator follows the pattern of exposing the underlying component's command line flags through CR field/property.

e.g.,

https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/apis/monitoring/v1/types.go#L159-L194
https://github.com/prometheus-operator/prometheus-operator/blob/master/pkg/apis/monitoring/v1/thanos_types.go#L107-L148

Though the existing model has better documentation & offers tight control on each supported cmd line flags., it loses out the flexibility when the [user wants to pass additional cmd args](https://github.com/openshift/thanos/pull/55#discussion_r620298796) which are not supported in CRs. 

Obviously user has the option to build their own image and configure it with desired default flags, but this is not for everyone.

**Comment by paulfantom** [comment]
Args can be overridden by using a strategic merge patch.

It is operator's job to prevent misconfiguration of the operand. The easiest way to achieve this is to allow only a specific set of arguments passed down to the operand and expose those options via CRD fields. Adding `extraArgs` field to the CRD removes the primary function of the operator as there is no way to correctly ensure args are correct without starting the binary. And if the operator needs to start operand just to check parameters, it already failed at its job.
